pydantic/types.py

```diff
@@ -505,10 +507,21 @@ def list_length_validator(cls, v: 'Optional[List[T]]') -> 'Optional[List[T]]':
 
         return v
 
+    @classmethod
+    def unique_items_validator(cls, v: 'Optional[List[T]]') -> 'Optional[List[T]]':
+        if cls.unique_items and v and len(set(v)) != len(v):
+            raise errors.ListUniqueItemsError(not_unique=len(v) - len(set(v)))
```

Comment:
you can avoid duplicating the `len(v)` and `len(set(v))` calls.

Reply:
What are we going to do for lists of unhashable types?

Maybe we just decide that is a limitation of this feature, but at the least we should document it.

Reply:
They are incidentally duplicated because the exception message is just a placeholder.

> Current error message is a placeholder. As specs do not state which mathematical object corresponds to this schema, there's no clearly defined concept of multiplicity. E.g. in `[1, 1, 2, 2, 2, 3]` one could understand: two items (1, 2), three items (1, 2, 2), or five items (1, 1, 2, 2, 2). I tend to see it as a set, so the first statement makes more sense to me.

I still don't know how many elements are supposed to be leftover if validation fails. Do you have an opinion on this? If not I can try to open an issue at the [JSON Schema repo](https://github.comjson-schema-org/json-schema-spec).

But in this case it would also be a micro-optimization, since the second calls are only made on exceptions, and this would save two `STORE_FAST` and two `LOAD_FAST` in a successful validation.

---

We can add a failover for unhashable types. E.g. something such as _(with the appropriate exceptions)_

```python
try:
    if v and len(set(v)) != len(v):
        raise ValueError
except TypeError:
    unique = list()
    if len([unique.append(i) for i in v if i not in unique]) != len(v):
        raise ValueError from None
```

I can benchmark some options. IMHO, as in the previous case, I think it is worth optimizing the check at the expense of a more expensive validation error. What do you think?

Reply:
json-schema-org/json-schema-spec#1099

Reply:
The failover for unhashable types seems too tricky to mypy 😅 Would you accept a `type: ignore` if it turns out to be the most performant?

Reply:
There's currently no instance-based output structure for validation errors (only schema-based). So, if it's okay with you, I'll continue with a generic error message to move on to the unhashable types validation.